File: pkg/templates/python/moondream-groq-computer-use/tools/computer.py
"""
Computer Tool - Maps high-level actions to Kernel Computer Controls API.
"""

import asyncio
import base64

from kernel import Kernel

from .types import (
    ComputerAction,
    ComputerFunctionArgs,
    PREDEFINED_COMPUTER_USE_FUNCTIONS,
    DEFAULT_SCREEN_SIZE,
    COORDINATE_SCALE,
    ToolResult,
    ScreenSize,
)


TYPING_DELAY_MS = 12
SCREENSHOT_DELAY_SECS = 0.5


class ComputerTool:
    def __init__(
        self,
        kernel: Kernel,
        session_id: str,
        screen_size: ScreenSize = DEFAULT_SCREEN_SIZE,
    ):
        self.kernel = kernel
        self.session_id = session_id
        self.screen_size = screen_size

    def denormalize_x(self, x: int) -> int:
        return int((x / COORDINATE_SCALE) * self.screen_size.width)

    def denormalize_y(self, y: int) -> int:
        return int((y / COORDINATE_SCALE) * self.screen_size.height)

    async def screenshot(self) -> ToolResult:
        try:
            await asyncio.sleep(SCREENSHOT_DELAY_SECS)
            response = self.kernel.browsers.computer.capture_screenshot(self.session_id)
            screenshot_bytes = response.read()
            dimensions = _parse_png_dimensions(screenshot_bytes)
            if dimensions:
                self.screen_size = dimensions

            return ToolResult(
                base64_image=base64.b64encode(screenshot_bytes).decode("utf-8"),
                url="about:blank",
                width=dimensions.width if dimensions else None,
                height=dimensions.height if dimensions else None,
            )
        except Exception as e:
            return ToolResult(error=f"Failed to take screenshot: {e}", url="about:blank")

    async def execute_action(
        self, action_name: str, args: ComputerFunctionArgs
    ) -> ToolResult:
        if action_name not in [a.value for a in PREDEFINED_COMPUTER_USE_FUNCTIONS]:
            return ToolResult(error=f"Unknown action: {action_name}")

        try:
            if action_name == ComputerAction.OPEN_WEB_BROWSER:
                # Browser is already open in Kernel, just return screenshot
                pass

            elif action_name == ComputerAction.CLICK_AT:
                if "x" not in args or "y" not in args:
                    return ToolResult(error="click_at requires x and y coordinates")
                x = self.denormalize_x(args["x"])
                y = self.denormalize_y(args["y"])
                num_clicks = int(args.get("clicks", 1)) if args.get("clicks") else 1
                self.kernel.browsers.computer.click_mouse(
                    self.session_id,
                    x=x,
                    y=y,
                    button="left",
                    click_type="click",
                    num_clicks=num_clicks,
                )

            elif action_name == ComputerAction.HOVER_AT:
                if "x" not in args or "y" not in args:
                    return ToolResult(error="hover_at requires x and y coordinates")
                x = self.denormalize_x(args["x"])
                y = self.denormalize_y(args["y"])
                self.kernel.browsers.computer.move_mouse(
                    self.session_id, x=x, y=y
                )

            elif action_name == ComputerAction.TYPE_TEXT_AT:
                if "x" not in args or "y" not in args:
                    return ToolResult(error="type_text_at requires x and y coordinates")
                if "text" not in args:
                    return ToolResult(error="type_text_at requires text")

                x = self.denormalize_x(args["x"])
                y = self.denormalize_y(args["y"])

                self.kernel.browsers.computer.click_mouse(
                    self.session_id,
                    x=x,
                    y=y,
                    button="left",
                    click_type="click",
                    num_clicks=1,
                )

                if args.get("clear_before_typing", True):
                    self.kernel.browsers.computer.press_key(
                        self.session_id, keys=["ctrl+a"]
                    )
                    await asyncio.sleep(0.05)

                self.kernel.browsers.computer.type_text(
                    self.session_id,
                    text=args["text"],
                    delay=TYPING_DELAY_MS,
                )

                if args.get("press_enter", False):
                    await asyncio.sleep(0.1)
                    self.kernel.browsers.computer.press_key(
                        self.session_id, keys=["Return"]
                    )

            elif action_name == ComputerAction.SCROLL_DOCUMENT:
                if "direction" not in args:
                    return ToolResult(error="scroll_document requires direction")
                center_x = self.screen_size.width // 2
                center_y = self.screen_size.height // 2
                scroll_delta = 500

                delta_x, delta_y = 0, 0
                direction = args["direction"]
                if direction == "down":
                    delta_y = scroll_delta
                elif direction == "up":
                    delta_y = -scroll_delta
                elif direction == "right":
                    delta_x = scroll_delta
                elif direction == "left":
                    delta_x = -scroll_delta

                self.kernel.browsers.computer.scroll(
                    self.session_id,
                    x=center_x,
                    y=center_y,
                    delta_x=delta_x,
                    delta_y=delta_y,
                )

            elif action_name == ComputerAction.SCROLL_AT:
                if "x" not in args or "y" not in args:
                    return ToolResult(error="scroll_at requires x and y coordinates")
                if "direction" not in args:
                    return ToolResult(error="scroll_at requires direction")

                x = self.denormalize_x(args["x"])
                y = self.denormalize_y(args["y"])

                magnitude = args.get("magnitude", 800)
                direction = args["direction"]
                if direction in ("up", "down"):
                    magnitude = self.denormalize_y(magnitude)
                else:
                    magnitude = self.denormalize_x(magnitude)

                delta_x, delta_y = 0, 0
                if direction == "down":
                    delta_y = magnitude
                elif direction == "up":
                    delta_y = -magnitude
                elif direction == "right":
                    delta_x = magnitude
                elif direction == "left":
                    delta_x = -magnitude

                self.kernel.browsers.computer.scroll(
                    self.session_id,
                    x=x,
                    y=y,
                    delta_x=delta_x,
                    delta_y=delta_y,
                )

            elif action_name == ComputerAction.WAIT_5_SECONDS:
                await asyncio.sleep(5)

            elif action_name == ComputerAction.GO_BACK:
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=["alt+Left"]
                )
                await asyncio.sleep(1)

            elif action_name == ComputerAction.GO_FORWARD:
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=["alt+Right"]
                )
                await asyncio.sleep(1)

            elif action_name == ComputerAction.SEARCH:
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=["ctrl+l"]
                )

            elif action_name == ComputerAction.NAVIGATE:
                if "url" not in args:
                    return ToolResult(error="navigate requires url")
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=["ctrl+l"]
                )
                await asyncio.sleep(0.1)
                self.kernel.browsers.computer.type_text(
                    self.session_id,
                    text=args["url"],
                    delay=TYPING_DELAY_MS,
                )
                await asyncio.sleep(0.1)
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=["Return"]
                )
                await asyncio.sleep(1.5)

            elif action_name == ComputerAction.KEY_COMBINATION:
                if "keys" not in args:
                    return ToolResult(error="key_combination requires keys")
                keys = str(args["keys"])
                if keys.lower() == "enter":
                    keys = "Return"
                self.kernel.browsers.computer.press_key(
                    self.session_id, keys=[keys]
                )

            elif action_name == ComputerAction.DRAG_AND_DROP:
                required = ["x", "y", "destination_x", "destination_y"]
                if not all(k in args for k in required):
                    return ToolResult(
                        error="drag_and_drop requires x, y, destination_x, and destination_y"
                    )

                start_x = self.denormalize_x(args["x"])
                start_y = self.denormalize_y(args["y"])
                end_x = self.denormalize_x(args["destination_x"])
                end_y = self.denormalize_y(args["destination_y"])

                self.kernel.browsers.computer.drag_mouse(
                    self.session_id,
                    path=[[start_x, start_y], [end_x, end_y]],
                    button="left",
                )

            else:
                return ToolResult(error=f"Unhandled action: {action_name}")

            await asyncio.sleep(SCREENSHOT_DELAY_SECS)
            return await self.screenshot()

        except Exception as e:
            return ToolResult(error=f"Action failed: {e}", url="about:blank")


def _parse_png_dimensions(data: bytes) -> ScreenSize | None:
    if len(data) < 24:
        return None
    if data[:8] != b"\x89PNG\r\n\x1a\n":
        return None
    width = int.from_bytes(data[16:20], "big")
    height = int.from_bytes(data[20:24], "big")
    if width <= 0 or height <= 0:
        return None
    return ScreenSize(width=width, height=height)

File: pkg/templates/typescript/moondream-groq-computer-use/tools/computer.ts
/**
 * Computer Tool - Maps high-level actions to Kernel's Computer Controls API.
 */

import { Buffer } from 'buffer';
import type { Kernel } from '@onkernel/sdk';
import {
  ComputerAction,
  PREDEFINED_COMPUTER_USE_FUNCTIONS,
  DEFAULT_SCREEN_SIZE,
  COORDINATE_SCALE,
  type ComputerFunctionArgs,
  type ToolResult,
  type ScreenSize,
} from './types/computer';

const TYPING_DELAY_MS = 12;
const SCREENSHOT_DELAY_MS = 500;

export class ComputerTool {
  private kernel: Kernel;
  private sessionId: string;
  private screenSize: ScreenSize;

  constructor(kernel: Kernel, sessionId: string, screenSize: ScreenSize = DEFAULT_SCREEN_SIZE) {
    this.kernel = kernel;
    this.sessionId = sessionId;
    this.screenSize = screenSize;
  }

  getScreenSize(): ScreenSize {
    return this.screenSize;
  }

  getKernel(): Kernel {
    return this.kernel;
  }

  getSessionId(): string {
    return this.sessionId;
  }

  private denormalizeX(x: number): number {
    return Math.round((x / COORDINATE_SCALE) * this.screenSize.width);
  }

  private denormalizeY(y: number): number {
    return Math.round((y / COORDINATE_SCALE) * this.screenSize.height);
  }

  async screenshot(): Promise<ToolResult> {
    try {
      await this.sleep(SCREENSHOT_DELAY_MS);
      const response = await this.kernel.browsers.computer.captureScreenshot(this.sessionId);
      const blob = await response.blob();
      const arrayBuffer = await blob.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);
      const dimensions = parsePngDimensions(buffer);
      if (dimensions) {
        this.screenSize = dimensions;
      }

      return {
        base64Image: buffer.toString('base64'),
        url: 'about:blank',
        width: dimensions?.width,
        height: dimensions?.height,
      };
    } catch (error) {
      return {
        error: `Failed to take screenshot: ${error}`,
        url: 'about:blank',
      };
    }
  }

  async executeAction(actionName: string, args: ComputerFunctionArgs): Promise<ToolResult> {
    if (!PREDEFINED_COMPUTER_USE_FUNCTIONS.includes(actionName as ComputerAction)) {
      return { error: `Unknown action: ${actionName}` };
    }

    try {
      switch (actionName) {
        case ComputerAction.OPEN_WEB_BROWSER:
          break;

        case ComputerAction.CLICK_AT: {
          if (args.x === undefined || args.y === undefined) {
            return { error: 'click_at requires x and y coordinates' };
          }
          const x = this.denormalizeX(args.x);
          const y = this.denormalizeY(args.y);
          const numClicks = typeof args.clicks === 'number' ? args.clicks : 1;
          await this.kernel.browsers.computer.clickMouse(this.sessionId, {
            x,
            y,
            button: 'left',
            click_type: 'click',
            num_clicks: numClicks,
          });
          break;
        }

        case ComputerAction.HOVER_AT: {
          if (args.x === undefined || args.y === undefined) {
            return { error: 'hover_at requires x and y coordinates' };
          }
          const x = this.denormalizeX(args.x);
          const y = this.denormalizeY(args.y);
          await this.kernel.browsers.computer.moveMouse(this.sessionId, { x, y });
          break;
        }

        case ComputerAction.TYPE_TEXT_AT: {
          if (args.x === undefined || args.y === undefined) {
            return { error: 'type_text_at requires x and y coordinates' };
          }
          if (!args.text) {
            return { error: 'type_text_at requires text' };
          }

          const x = this.denormalizeX(args.x);
          const y = this.denormalizeY(args.y);

          await this.kernel.browsers.computer.clickMouse(this.sessionId, {
            x,
            y,
            button: 'left',
            click_type: 'click',
            num_clicks: 1,
          });

          if (args.clear_before_typing !== false) {
            await this.kernel.browsers.computer.pressKey(this.sessionId, {
              keys: ['ctrl+a'],
            });
            await this.sleep(50);
          }

          await this.kernel.browsers.computer.typeText(this.sessionId, {
            text: args.text,
            delay: TYPING_DELAY_MS,
          });

          if (args.press_enter) {
            await this.sleep(100);
            await this.kernel.browsers.computer.pressKey(this.sessionId, {
              keys: ['Return'],
            });
          }
          break;
        }

        case ComputerAction.SCROLL_DOCUMENT: {
          if (!args.direction) {
            return { error: 'scroll_document requires direction' };
          }
          const centerX = Math.round(this.screenSize.width / 2);
          const centerY = Math.round(this.screenSize.height / 2);
          const scrollDelta = 500;

          let deltaX = 0;
          let deltaY = 0;
          if (args.direction === 'down') deltaY = scrollDelta;
          else if (args.direction === 'up') deltaY = -scrollDelta;
          else if (args.direction === 'right') deltaX = scrollDelta;
          else if (args.direction === 'left') deltaX = -scrollDelta;

          await this.kernel.browsers.computer.scroll(this.sessionId, {
            x: centerX,
            y: centerY,
            delta_x: deltaX,
            delta_y: deltaY,
          });
          break;
        }

        case ComputerAction.SCROLL_AT: {
          if (args.x === undefined || args.y === undefined) {
            return { error: 'scroll_at requires x and y coordinates' };
          }
          if (!args.direction) {
            return { error: 'scroll_at requires direction' };
          }

          const x = this.denormalizeX(args.x);
          const y = this.denormalizeY(args.y);

          let magnitude = args.magnitude ?? 800;
          if (args.direction === 'up' || args.direction === 'down') {
            magnitude = this.denormalizeY(magnitude);
          } else {
            magnitude = this.denormalizeX(magnitude);
          }

          let deltaX = 0;
          let deltaY = 0;
          if (args.direction === 'down') deltaY = magnitude;
          else if (args.direction === 'up') deltaY = -magnitude;
          else if (args.direction === 'right') deltaX = magnitude;
          else if (args.direction === 'left') deltaX = -magnitude;

          await this.kernel.browsers.computer.scroll(this.sessionId, {
            x,
            y,
            delta_x: deltaX,
            delta_y: deltaY,
          });
          break;
        }

        case ComputerAction.WAIT_5_SECONDS:
          await this.sleep(5000);
          break;

        case ComputerAction.GO_BACK:
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: ['alt+Left'],
          });
          await this.sleep(1000);
          break;

        case ComputerAction.GO_FORWARD:
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: ['alt+Right'],
          });
          await this.sleep(1000);
          break;

        case ComputerAction.SEARCH:
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: ['ctrl+l'],
          });
          break;

        case ComputerAction.NAVIGATE: {
          if (!args.url) {
            return { error: 'navigate requires url' };
          }
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: ['ctrl+l'],
          });
          await this.sleep(100);
          await this.kernel.browsers.computer.typeText(this.sessionId, {
            text: args.url,
            delay: TYPING_DELAY_MS,
          });
          await this.sleep(100);
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: ['Return'],
          });
          await this.sleep(1500);
          break;
        }

        case ComputerAction.KEY_COMBINATION: {
          if (!args.keys) {
            return { error: 'key_combination requires keys' };
          }
          const keyValue = String(args.keys).toLowerCase() === 'enter' ? 'Return' : args.keys;
          await this.kernel.browsers.computer.pressKey(this.sessionId, {
            keys: [keyValue],
          });
          break;
        }

        case ComputerAction.DRAG_AND_DROP: {
          if (args.x === undefined || args.y === undefined ||
              args.destination_x === undefined || args.destination_y === undefined) {
            return { error: 'drag_and_drop requires x, y, destination_x, and destination_y' };
          }

          const startX = this.denormalizeX(args.x);
          const startY = this.denormalizeY(args.y);
          const endX = this.denormalizeX(args.destination_x);
          const endY = this.denormalizeY(args.destination_y);

          await this.kernel.browsers.computer.dragMouse(this.sessionId, {
            path: [[startX, startY], [endX, endY]],
            button: 'left',
          });
          break;
        }

        default:
          return { error: `Unhandled action: ${actionName}` };
      }

      await this.sleep(SCREENSHOT_DELAY_MS);
      return await this.screenshot();

    } catch (error) {
      return { error: `Action failed: ${error}`, url: 'about:blank' };
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

function parsePngDimensions(buffer: Buffer): ScreenSize | null {
  if (buffer.length < 24) return null;
  const signature = buffer.subarray(0, 8);
  const expected = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
  if (!signature.equals(expected)) return null;
  const width = buffer.readUInt32BE(16);
  const height = buffer.readUInt32BE(20);
  if (!width || !height) return null;
  return { width, height };
}
